### ISSUE TITLE
gtk imshow deadlock

### ISSUE THREAD

**Issue by huangqinjin** [opencv/opencv#9002]
##### System information (version)

- OpenCV => 3.2
- Operating System / Platform => ubuntu 16.04
- Compiler => gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.4)

##### Detailed description

`opencv/modules/highgui/src/window_gtk.cpp`:
`cvShowImage` holds the `window_mutex` and then calls `cvNamedWindow`.  That would produce a deadlock on `window_mutex`.

##### Steps to reproduce

```.cpp
    // C++ code example
    cv::startWindowThread();
    // cv::namedWindow("img");  // ok if uncomment this
    cv::imshow("img", img);
 ```

**Comment by kunaltyagi** [comment]
##### Possibly related issue
On Ubuntu 18.04, using pre-puilt binary, there is an error with how the mutex is used. OpenCV (v3.2) is built with GTK (v3.0) as per `ldd`.
````c++
#include <opencv2/highgui.hpp>
int main() {
    cv::startWindowThread();
    cv::waitKey(0);
}
````
##### Error
```
Attempt to unlock mutex that was not locked
Aborted (core dumped)
```
Is this a related issue or a completely different one?

**Comment by alalek** [comment]
Thank you for the report!
Could you take a look on this fix: #11903 ?
